### PR TITLE
Use net/url to escape paths in web-ui

### LINKED ISSUE #2061: gateway: URI-escaping of paths
For example, if file url's contain @ or # they don't work

`https://ipfs.io/ipfs/QmZbfX4p7LUtuWgLtp1XYt6tkPnwfhqBTzqPDCanucYQrs/@#$_%5e_%20chillmix.mp3`

will return
`Path Resolve error: no link named "@" under QmZbfX4p7LUtuWgLtp1XYt6tkPnwfhqBTzqPDCanucYQrs`

Or

`https://ipfs.io/ipfs/QmZbfX4p7LUtuWgLtp1XYt6tkPnwfhqBTzqPDCanucYQrs/Cat%20D###%20Goo.mp3`

will return
`Path Resolve error: no link named "Cat D" under QmZbfX4p7LUtuWgLtp1XYt6tkPnwfhqBTzqPDCanucYQrs`

Directory
`https://ipfs.io/ipfs/QmZbfX4p7LUtuWgLtp1XYt6tkPnwfhqBTzqPDCanucYQrs/`

Tested on chrome and firefox

### PR DESCRIPTION
**Pull request by adrian-bl**
This change uses net/url to escape href's in the gateway directory listing instead of the builtin escaper by html/template which does not escape fragments (#) and other characters such as '?'.

Example:
```
$ wget -q -O - "http://localhost:8080/ipfs/QmcoMbJVCufZQEs1k3ez56JhHKTK9Npwo5UNmsQiwz7NYg/tis%23is/a'strange%3Fpath" | grep file
            <a href="/ipfs/QmcoMbJVCufZQEs1k3ez56JhHKTK9Npwo5UNmsQiwz7NYg/tis%23is/a%27strange%3Fpath/with%23one-filey%3F.txt">with#one-filey?.txt</a>
```
Fixes #2061

Note: I'm not sure about what to do with the vendor dir: should i re-publish it to ipfs and use the new hash? i was unable to find (and re-construct, using ipfs add -r) the currently used hash.

### CODE REVIEW

**Comment by whyrusleeping** [comment]
@adrian-bl for the license signoff thing, you can just amend your commit and force push (saves notification noise).

Also, don't worry about the vendor directory stuff. Thats all going to be 'fixed' after 0.4.0 with gx, so i'll handle it all then.

**Comment by adrian-bl** [comment]
@whyrusleeping ok, thanks. I already configured a git hook, so it shouldn't happen anymore anyways :-)

File: core/corehttp/gateway_indexPage.go [hunk 1]
```diff
@@ -45,6 +46,12 @@ func init() {
 		return "ipfs-" + ext[1:] // slice of the first dot
 	}
 
+	// custom template-escaping function to escape a full path, including '#' and '?'
+	urlEscape := func(rawUrl string) string {
+		pathUrl := url.URL{Path: rawUrl}
+		return pathUrl.String()
```
Comment (jbenet) [inline core/corehttp/gateway_indexPage.go]: this looks right to me

**Comment by jbenet** [comment]
@adrian-bl this LGTM! thanks! 

i think we want a test to go in with this PR though to ensure it's not broken in the future.

- looks like the gateway testing stuff is here: https://github.com/ipfs/go-ipfs/blob/master/core/corehttp/gateway_test.go
- though not sure where the directory listing stuff is tested, if at all. 

cc @cryptix as he knows this area better than me.

**Comment by lgierth** [comment]
> though not sure where the directory listing stuff is tested, if at all. 

Same file :) The "backlinks" related code

**Comment by lgierth** [comment]
Does this also fix request URIs containing escaped characters, as in https://github.com/ipfs/go-ipfs/issues/2061#issuecomment-164259299?

**Comment by adrian-bl** [comment]
@lgierth Yes, as net/url escapes the whole path according to RFC 3986, so the percent sign itself will be escaped to '%25':
```
$ echo bar >  'funky%20named+file.txt'
$ ipfs add -r .
added QmYNmQKp6SuaVrpgWRsPTgCQCnpxUYGq76YEKBXuj2N4H6 foo/funky%20named+file.txt
added QmTU5nJJEHaGSjDgqtawBSVUWmDbx9PPkoExaZbFkDNHxm foo
$ wget -O - http://localhost:8080/ipfs/QmTU5nJJEHaGSjDgqtawBSVUWmDbx9PPkoExaZbFkDNHxm|grep named
<a href="/ipfs/QmTU5nJJEHaGSjDgqtawBSVUWmDbx9PPkoExaZbFkDNHxm/funky%2520named&#43;file.txt">funky%20named&#43;file.txt</a>
```

**Comment by rht** [comment]
I just verified this pr by testing the directory listing directly.
@adrian-bl you should include the test file from the gist.
This needs a rebase.

This can be added to sharness either here or in dir-index-html repo
```
test_expect_success "gateway dir listing escapes the characters" '
    mkdir -p ttt &&
    echo foo >ttt/@\#\#_\? &&
    ipfs add -r ttt &&
    curl 'localhost:8080/ipfs/QmantKYPCRwwifWxr3DcrqrStqULS4vJ7x1n9R8B4gCSrJ/' >curl_out &&
    grep "QmantKYPCRwwifWxr3DcrqrStqULS4vJ7x1n9R8B4gCSrJ/@%23%23_%3F" curl_out
'
```

**Comment by jbenet** [comment]
@rht thanks for resurfacing this + the review. want to pick this up and rebase the PR on top of master + add test?

**Comment by whyrusleeping** [comment]
ping @noffle, think you could pick this up?

**Pull request closed without merging**